dashboard/lib/services/lti/account_linker.rb

```diff
@@ -12,13 +12,40 @@ def call
         ActiveRecord::Base.transaction do
           user.authentication_options << rehydrated_user.authentication_options.first
           Services::Lti.create_lti_user_identity(user)
+          handle_sections(rehydrated_user, user)
           user.lti_roster_sync_enabled = true if user.teacher?
           PartialRegistration.delete(session)
+          rehydrated_user.destroy if rehydrated_user.id
         end
       end
 
       private def rehydrated_user
-        @rehydrated_user ||= ::User.new_with_session(ActionController::Parameters.new, session)
+        @rehydrated_user ||= find_cached_user || ::User.new_with_session(ActionController::Parameters.new, session)
+      end
+
+      # The new user might already be in sections, if the account was created
+      # via a roster sync. In this case, we need to swap the pre-existing user
+      # into these sections to avoid the roster being in a bad state.
+      private def handle_sections(user_to_remove, user_to_add)
+        return if user_to_remove.sections_as_student.empty?
+        user_to_remove.sections_as_student.each do |section|
+          section.students << user_to_add
+          section.students.destroy(user_to_remove)
```

Comment:
Looks like there's a test failure here:

```
=ERROR["test_Swaps_the_existing_user_into_the_defunct_user's_sections", "Services::Lti::AccountLinkerTest", 1075.3661866319999]
 test_Swaps_the_existing_user_into_the_defunct_user's_sections#Services::Lti::AccountLinkerTest (1075.37s)
Minitest::UnexpectedError:         ArgumentError: You are passing an instance of ActiveRecord::Base to `find`. Please pass the id of the object by calling `.id`.
            lib/services/lti/account_linker.rb:33:in `block in handle_sections'
            lib/services/lti/account_linker.rb:31:in `handle_sections'
            lib/services/lti/account_linker.rb:15:in `block in call'
            lib/services/lti/account_linker.rb:12:in `call'
            lib/services/base.rb:4:in `call'
            test/lib/services/lti/account_linker_test.rb:42:in `block in <class:AccountLinkerTest>'
            test/testing/setup_all_and_teardown_all.rb:36:in `run'
```

Reply:
Not sure exactly what was causing this, but I couldn't repro this locally. Re-running the tests in drone caused other failures, but after rebasing post-mysql-8 upgrade, the tests all pass.